apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/sampling/SamplingService.java

```diff
@@ -114,4 +113,26 @@ public void forceSampled() {
     private void resetSamplingFactor() {
         samplingFactorHolder = new AtomicInteger(0);
     }
+
+    /**
+     * Reset scheduledFuture to support samplingRate changed.
+     */
+    void resetScheduledFuture() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+        }
+        on = false;
+        if (samplingRateWatcher.getSamplingRate() > 0) {
```

Comment:
You are still checking and reading this number repeatedly. Add `boolean shouldSampling` to the `samplingRateWatcher`